File: ios/ReactNativeCodex/RuntimeDiagnosticsModule.swift
import Foundation
import React
import UIKit

@objc(RuntimeDiagnosticsModule)
class RuntimeDiagnosticsModule: NSObject {
  @objc
  static func requiresMainQueueSetup() -> Bool {
    false
  }

  @objc(getRuntimeProfile:rejecter:)
  func getRuntimeProfile(
    _ resolve: RCTPromiseResolveBlock,
    rejecter reject: RCTPromiseRejectBlock
  ) {
    let processInfo = ProcessInfo.processInfo
    let thermalState = processInfo.thermalState
    let lowPowerModeEnabled = processInfo.isLowPowerModeEnabled

    let performanceTier: String
    switch thermalState {
    case .critical:
      performanceTier = "critical"
    case .serious, .fair:
      performanceTier = "constrained"
    case .nominal:
      performanceTier = lowPowerModeEnabled ? "constrained" : "standard"
    @unknown default:
      performanceTier = lowPowerModeEnabled ? "constrained" : "standard"
    }

    let appVersion =
      Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String ?? "unknown"

    resolve([
      "platform": "ios",
      "operatingSystemVersion": UIDevice.current.systemVersion,
      "deviceModel": UIDevice.current.model,
      "appVersion": appVersion,
      "lowPowerModeEnabled": lowPowerModeEnabled,
      "performanceTier": performanceTier,
      "diagnosticsSource": "native-module",
    ])
  }
}

File: android/app/src/main/java/com/reactnativecodex/nativeintegration/RuntimeDiagnosticsModule.kt
package com.reactnativecodex.nativeintegration

import android.app.ActivityManager
import android.content.Context
import android.os.Build
import android.os.PowerManager
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod

class RuntimeDiagnosticsModule(
  private val reactContext: ReactApplicationContext,
) : ReactContextBaseJavaModule(reactContext) {

  override fun getName(): String = "RuntimeDiagnosticsModule"

  @ReactMethod
  fun getRuntimeProfile(promise: Promise) {
    try {
      val activityManager =
        reactContext.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
      val powerManager =
        reactContext.getSystemService(Context.POWER_SERVICE) as PowerManager

      val memoryClass = activityManager.memoryClass
      val lowRamDevice = activityManager.isLowRamDevice
      val lowPowerMode = powerManager.isPowerSaveMode

      val performanceTier =
        when {
          lowRamDevice && lowPowerMode -> "critical"
          lowRamDevice || lowPowerMode || memoryClass <= 192 -> "constrained"
          else -> "standard"
        }

      val packageInfo =
        reactContext.packageManager.getPackageInfo(reactContext.packageName, 0)
      val appVersion = packageInfo.versionName ?: "unknown"

      val payload =
        Arguments.createMap().apply {
          putString("platform", "android")
          putString("operatingSystemVersion", "${Build.VERSION.RELEASE} (API ${Build.VERSION.SDK_INT})")
          putString("deviceModel", "${Build.MANUFACTURER} ${Build.MODEL}")
          putString("appVersion", appVersion)
          putBoolean("lowPowerModeEnabled", lowPowerMode)
          putString("performanceTier", performanceTier)
          putString("diagnosticsSource", "native-module")
        }

      promise.resolve(payload)
    } catch (error: Exception) {
      promise.reject("RUNTIME_PROFILE_ERROR", error.message, error)
    }
  }
}
